#This R script performs the following operations
#1) read the read.fasta file in the current working directory
#2) cluster the reads according to k-mer frequency
#   look for see "kmer_sz = " to see what k-mer length is being used
#3) cluster using the best cluster size for k-means, which
#   is done by eye using the plot generated by plot(explainedVar)
#   command
#4) run velvet on each cluster (in parallel)
#5) for each set of contigs in each cluster
#   
#

library("seqinr")
library("cluster")
library("fpc")
library("parallel")

#STEP 1: load fasta and count k-mers
fas <- read.fasta(file = "reads.fasta")
n = length(fas)

#STEP 2: kmer frequencies
kmer_sz = 2
kmers = matrix(0, n, 4^kmer_sz)
for (i in 1:n) {
  kmers[i,] = count(fas[[i]], kmer_sz)
}

#STEP 3: Identify the best cluster size
koptions = c(5,10,20,30,50,80,100,200)
n = length(koptions)
explainedVar = matrix(0,n,2)
for (i in 1:n) {
  clusters = kmeans( kmers, koptions[i] , 100);
  explainedVar[i,1] = koptions[i]
  explainedVar[i,2] = clusters$betweenss/clusters$totss
}
plot(explainedVar)

#STEP 3 cont'd: Final clustering
numClusters = 50
finalClusters = kmeans( kmers, numClusters , 100)

#STEP 4: Group FASTA sequences
fastaSeqs = list()
fastaNames = list()
for (i in 1:numClusters) {
  fastaSeqs[[i]] = list()
  fastaNames[[i]] = list()
}

n = length(fas)
for (i in 1:n) {
  cluster = finalClusters$cluster[i]
  m = length(fastaSeqs[[cluster]])
  fastaSeqs[[cluster]][[m+1]] = fas[[i]]
  fastaNames[[cluster]][[m+1]] = attr(fas[[i]], "name")
}


##STEP 4 cont'd: for each cluster, create contigs
velvetcmd = "perl"
velvetargs = "~/VelvetOptimiser/VelvetOptimiser.pl -t 2 -s 27 -e 41 -f '-short -fasta ~/temp.fasta'"
pwd = getwd()

#For parallel proc
callVelvetPar = function(i, pwd, fastaSeqs, fastaNames, velvetcmd, velvetargs) {

  library("seqinr")

  wd = paste(pwd, "/iter_", i, sep="")
  system2("mkdir", wd)
  setwd(wd)

  write.fasta(fastaSeqs[[i]], fastaNames[[i]], "temp.fasta")
  ret = system2(velvetcmd, velvetargs, stdout=TRUE, stderr=TRUE)
  j = 0
  for (i in 1:length(ret)) {
    if (ret[i] == "Assembly output files are in the following directory:") {
      j = i + 1
      break;
    }
  }
  if (j > 0) {
    contig_file = paste(ret[j],"/contigs.fa", sep="")
    print(contig_file)
  }
}

cl = makeCluster(4, outfile="/tmp/output")
contigs = parLapply(cl, 1:numClusters, callVelvetPar, pwd, fastaSeqs, fastaNames, velvetcmd, velvetargs)
stopCluster(cl)

save.image(".RData")

for (i in 1:length(contigs)) { 
  print(contigs[[i]]); 
}

